### ISSUE TITLE
Haskell: support cross-compiling TemplateHaskell with GHC via -fexternal-interpreter

### ISSUE THREAD

**Issue by sternenseemann** [NixOS/nixpkgs#248979]
TemplateHaskell notoriously requires running host code even when cross-compiling. This is realised by telling GHC about a so called iserv via the `-fexternal-interpreter` flag. This iserv can either run on a physical or virtual machine matching the host platform or run via qemu (user) emulation. We should allow for the latter in a convenient way in nixpkgs. `haskell.nix` already supports this and relies on the nixpkgs cross infrastructure, so we can probably compare notes.

**Comment by sternenseemann** [comment]
I seem to remember @amjoseph-nixpkgs did some work towards this.

**Comment by amjoseph-nixpkgs** [comment]
Yeah, if you use `-fexternal-interpreter=qemu-user` and try cross with `build=x86_64-linux`, `host=aarch64-linux` and you'll get a segfault deep within the bowels of GHC.  Turns out they implemented their own custom dynamic linker (?!?! why !?!?!?! why why why???) which can't cope with qemu's slightly-not-exactly-like-linux memory layout.

There's a bug on the GHC bugtracker with lots of details.  I found it by searching for the failure I was seeing.

**Comment by amjoseph-nixpkgs** [comment]
This is the most enlightening comment I encountered: https://gitlab.haskell.org/ghc/ghc/-/issues/21767#note_441415

That bug describes basically the final obstacle I hit.

**Comment by alexfmpe** [comment]
I think this can be closed right? 
There's lots of build->host combinations that'll need tweaks, but the core cross-compiling TH feature is merged now
